js_modules/dagit/packages/core/src/graph/PipelineGraph.tsx

```diff
@@ -176,28 +178,38 @@ class PipelineGraphContents extends React.PureComponent<
             strokeWidth={2}
           />
         ))}
-        <foreignObject width={layout.width} height={layout.height}>
-          {ops.map((op) => (
-            <OpNode
-              key={op.name}
-              invocation={op}
-              definition={op.definition}
-              minified={minified}
-              onClick={() => onClickOp({name: op.name})}
-              onDoubleClick={() => onDoubleClickOp({name: op.name})}
-              onEnterComposite={() => onEnterSubgraph({name: op.name})}
-              onHighlightEdges={this.onHighlightEdges}
-              layout={layout.ops[op.name]}
-              selected={selectedOp === op}
-              focused={focusOps.includes(op)}
-              highlightedEdges={
-                isOpHighlighted(this.state.highlighted, op.name)
-                  ? this.state.highlighted
-                  : EmptyHighlightedArray
-              }
-              dim={highlightedOps.length > 0 && highlightedOps.indexOf(op) === -1}
-            />
-          ))}
+        <foreignObject width={layout.width} height={layout.height} style={{pointerEvents: 'none'}}>
+          {ops
+            .filter((op) => {
+              const box = layout.ops[op.name].boundingBox;
+              return (
+                box.x + box.width >= bounds.left &&
+                box.y + box.height >= bounds.top &&
+                box.x < bounds.right &&
+                box.y < bounds.bottom
+              );
+            })
```

Comment:
I could see this being worth a `React.useMemo` if `PipelineGraphContents` were turned into an SFC.

Reply:
Yeah i think that'd make sense 👍  might be a good thing to try next. Now that we're passing in the bounding box of the viewport the PipelineGraphContents has to re-render as you pan / zoom so i'm not sure a pure component would skip many renders though. Right now it's just it's children that are memo'd.